lib/extensions/virtual_total.rb

```diff
@@ -93,17 +93,34 @@ def define_virtual_aggregate_method(name, relation, method_name, column)
     def virtual_aggregate_arel(reflection, method_name, column)
       return unless reflection && reflection.macro == :has_many && !reflection.options[:through]
       lambda do |t|
+        query = if reflection.scope
+                  reflection.klass.instance_exec(nil, &reflection.scope)
+                else
+                  reflection.klass.all
+                end
+
+        # ordering will probably screw up aggregations, so clear this out from
+        # any calls
+        #
+        # only clear this out if this isn't a `:size` call as well, since doing
+        # a COUNT(*) will allow any ORDER BY to still work properly.  This is
+        # to avoid any possible edge cases by clearing out the order clause.
+        query.order_values = [] if method_name != :size
+
         foreign_table = reflection.klass.arel_table
         # need db access for the keys, so delaying all this lookup until call time
-        local_key = reflection.active_record_primary_key
+        local_key   = reflection.active_record_primary_key
         foreign_key = reflection.foreign_key
+        query       = query.where(t[local_key].eq(foreign_table[foreign_key]))
+
         arel_column = if method_name == :size
                         Arel.star.count
                       else
                         reflection.klass.arel_attribute(column).send(method_name)
                       end
-        t.grouping(foreign_table.project(arel_column)
-                                .where(t[local_key].eq(foreign_table[foreign_key])))
+        query       = query.select(arel_column)
+
+        t.grouping(Arel::Nodes::SqlLiteral.new(query.to_sql))
```

Comment:
This is a big change, so I figure I should go through it (even if I have already).

Background info
---------------

`arel`, specifically with `ActiveRecord`, will handled params passed into `.where` clauses (and others), as `bind_params` (aka `bound_attributes`/`binds`... it has many names in the rails codebase), and they will be kept in the `where_clause_values` hash and applied when the SQL string is generated (either when executing the query, or when calling `.to_sql`).  You can see this in practice here:

https://github.com/rails/rails/blob/d8bc096/activerecord/lib/active_record/relation.rb#L598-L614

That said, there is actually no other way within the `arel` API to apply the bind attributes to an existing `ast` (abstract syntax tree) or`arel` object.  Instead, everything available in the API just converts to the binds, along with the rest of the query, into a SQL query string to be sent over the wire.  Further more, the binds can't really be passed from inside the `virtual_attribute` proc to the outer query calling it (from what I can tell).

To further illustrate what that would look like if we just left it as is, given the following `ActiveRecord` query:

```ruby
Vm.where(:smart => false)
```

Converting it to SQL without the binds converted would looks something like this:

```SQL
SELECT *
FROM vms
WHERE smart = $1
```

When it really should be:

```SQL
SELECT *
FROM vms
WHERE smart = 'f'
```

The exception is when it is something like `Vm.where(:ems_id => nil)`, since `nil` is a special value in the `AST` walker that will always convert to `IS NULL` (or `IS NOT NULL` if inverted), but every other value is assumed to be a bind.


Implementation
--------------
So given the above, there really was only two options available:

* write my own AST walker to convert existing `BindParams` in the tree.  This would have ended up a hack around the existing `visitor`/`collector` pattern in `arel`, since it is based mostly around walking that tree, and appending to a `@parts` array for each node.
* use the existing string generated `collector` (aka `.to_sql`), and wrap it in a `Arel::Nodes::SqlLiteral.new(query.to_sql)` to make it a valid AST.

I obviously choose the second, and this solves the issue that brought up this bug, but it has the side affect of making most of what is returned pretty much immutable instead of a flexible and walkable AST.

The other thing that was done in doing this was flipping the generation of the `virtual_attribute` SQL on it's head, and starting with the existing query and adding to it, instead of the previous implementation, which tried to build it from scratch.  Again, because of the binds, there isn't a clean and abstract way of getting this info from the query, so starting with what was there and building on top of it was the only way to allow the existing scopes on relations to function.  Trying to work this up by hand and taking from the existing ends up being very bulky and not really reusable (@lpichler 's PR is an example of [this implementation](https://github.com/ManageIQ/manageiq/pull/16512/files#diff-8c31e031a75a90caeae44d6b43300e2cR11))

The final bit, is because we are starting from an `ActiveRecord` query from the `scope`/`default_scope` of the relation, and not a pure `arel` query, we need to convert the `ActiveRecord` query into something that `arel` will accept (of note:  you can put `arel` in a `ActiveRecord` query, but not an `ActiveRecord` query in a `arel` query).  This is where the `Arel::Nodes::SqlLiteral.new(query.to_sql)` comes into play.

As mentioned previously, this makes the final product much more rigid, though I am unaware of any situation in MIQ where the internal `arel` of a `virtual_attribute` is being manipulated after it has been applied to a query, so this shouldn't be the biggest deal in the world.

Ideally there shouldn't be a huge affect with this change, but it does have the risk of affecting every `virtual_total`/`virtual_aggregation` in the application, which thanks to @kbrock ( :wink: ), is a non-trivial amount.